atc/engine/builder/builder.go

```diff
@@ -43,23 +46,35 @@ func NewStepBuilder(
 	redactSecrets bool,
 ) *stepBuilder {
 	return &stepBuilder{
-		stepFactory:     stepFactory,
-		delegateFactory: delegateFactory,
-		externalURL:     externalURL,
-		secrets:         secrets,
-		redactSecrets:   redactSecrets,
+		stepFactory:      stepFactory,
+		delegateFactory:  delegateFactory,
+		externalURL:      externalURL,
+		globalSecrets:    secrets,
+		redactSecrets:    redactSecrets,
+		pipelineCredMgrs: map[int][]creds.Manager{},
 	}
 }
 
 type stepBuilder struct {
-	stepFactory     StepFactory
-	delegateFactory DelegateFactory
-	externalURL     string
-	secrets         creds.Secrets
-	redactSecrets   bool
+	stepFactory      StepFactory
+	delegateFactory  DelegateFactory
+	externalURL      string
+	globalSecrets    creds.Secrets
+	redactSecrets    bool
+	pipelineCredMgrs map[int][]creds.Manager
 }
 
-func (builder *stepBuilder) BuildStep(build db.Build) (exec.Step, error) {
+func (builder *stepBuilder) FinishStep(build db.Build, logger lager.Logger) {
+	if builder.pipelineCredMgrs[build.ID()] != nil {
+		logger.Debug("close-pipeline-credential-managers")
+		for _, cm := range builder.pipelineCredMgrs[build.ID()] {
+			cm.Close(logger)
```

Comment:
I think pipeline is a static thing, a pipeline may not have any build for long time, it makes not much sense to maintain credential managers for no-build pipeline.

A credential manager just launches a go-routine, a build already has a lot of go-routines, so one more seems not much matter.

Reply:
Spending one goroutine is fine, but there's probably a heavier cost than just a single goroutine, since these goroutines create connections and that may involve doing a TLS handshake repeatedly if it's a fresh client every time.

Users have observed that constantly doing TLS handshakes can actually be very costly when talking to Postgres; increasing the idle connection pool size significantly reduced CPU usage on the `web` nodes because they didn't have to do the TLS handshake as often.

In looking at our Vault code, we use Vault with its default 'pooled' client, which will probably help mitigate the TLS handshake cost by re-using connections, but it actually warns that it shouldn't be used for 'transient clients':

https://github.com/hashicorp/vault/blob/ad3f31d7954a1c46b2c6e01f0c72a033ffd1117a/api/client.go#L142

https://github.com/hashicorp/vault/blob/ad3f31d7954a1c46b2c6e01f0c72a033ffd1117a/vendor/github.com/hashicorp/go-cleanhttp/cleanhttp.go#L49-L52

Now, this *might* not be a big deal because we'll likely be re-using the pooled client for a stable-ish set of hosts over time, but it's still a little worrying.

We don't have to solve this right off the bat, since it might be OK, but I could see it behaving poorly at medium-to-large scale. Maybe we just need to test it out at scale and see where it bends and/or breaks. :thinking:

Reply:
@evanchaoli Actually, won't we have to use these pipeline credential managers for resource checking, too? At that point we would basically always have an active client. Is that not implemented in the PR yet?

Reply:
> @evanchaoli Actually, won't we have to use these pipeline credential managers for resource checking, too? At that point we would basically always have an active client. Is that not implemented in the PR yet?

It's not implemented yet in this PR. Actually this is one thing I was not sure, I added the question at https://github.com/concourse/concourse/pull/4600/files#diff-e6d6a569fec760fc128319f5e4eda027R161. `check` step only interpolates variables for resource configure, right?

Reply:
Yep!

Reply:
@vito I checked `check` related code, now I agree pipeline credential managers should be maintained at pipeline level. Given multiple pipelines might use same credential manager, I'm thinking to create a global credential manager pool (map), using interpolated `var-source-config` as key, so that we can avoid duplicate credential manager instances. One question, what's plan of making `lidar` as default? In other words, do we need to support this feature for  non-lidar?

Reply:
@evanchaoli We plan to make `lidar` the default ASAP (based on user feedback as we get more mileage on it), at least by the time we ship v6.0, so I'm OK with not supporting non-lidar (aka radar).

A global credential manager pool makes sense to me! Would you be trying @pivotal-jwinters's suggestion of a `credential_managers` table?

FWIW you'll probably want to key on a hash of the config, similar to how we key resource configs.

There's probably more to figure out than I'm able to cover at this point, but we can figure it out as we go. :)

Reply:
I have added a `VarSourcePool` in `creds/pool.go`, it maintains a map of credential managers, key is config of managers. I have made both lidar and non-lidar work now.

Reply:
Yes, I have done keying on the config in the commit I just pushed. Also I have made both lidar and non-lidar `check`s work.